trio/_highlevel_pipes.py

```diff
@@ -0,0 +1,63 @@
+import os
+from typing import Tuple
+
+from . import _core
+from ._abc import SendStream, ReceiveStream
+from ._highlevel_socket import _translate_socket_errors_to_stream_errors
+
+__all__ = ["PipeSendStream", "PipeReceiveStream", "make_pipe"]
+
+
+class _PipeMixin:
+    def __init__(self, pipefd: int):
+        if not isinstance(pipefd, int):
+            raise TypeError("PipeSendStream needs a pipe fd")
+
+        self._pipe = pipefd
+
+    async def aclose(self):
+        os.close(self._pipe)
+
+    def fileno(self) -> int:
+        """Gets the file descriptor for this pipe."""
+        return self._pipe
+
+
+class PipeSendStream(_PipeMixin, SendStream):
+    """Represents a send stream over an os.pipe object."""
+
+    async def send_all(self, data: bytes):
+        length = len(data)
+        # adapted from the SocketStream code
+        with memoryview(data) as view, \
+                _translate_socket_errors_to_stream_errors():
+            total_sent = 0
+            while total_sent < length:
+                with view[total_sent:] as remaining:
+                    total_sent += os.write(self._pipe, remaining)
+
+                await self.wait_send_all_might_not_block()
```

Comment:
This is going to crash if the pipe buffer fills up – `os.write` will raise `BlockingIOError` (EWOULDBLOCK), and we don't have anything to catch it. In `SocketStream.send_all`, it has a slightly easier job, because it doesn't call the raw OS `send` function directly – it calls the `trio.socket` `send` method, and that takes care or retrying and waiting. See `_nonblocking_helper` and `_try_sync` in `trio.socket` for the gory details... not that they're perfect, they're probably a bit over-elaborate themselves.